src/lib/uptime_service/uptime_service.ml

```diff
@@ -0,0 +1,431 @@
+                                best_tip
+                            in
+                            let block_data =
+                              { block= block_base64
+                              ; created_at= get_rfc3339_time ()
+                              ; peer_id
+                              ; snark_work= Some snark_work_base64 }
+                            in
+                            send_uptime_data ~logger ~submitter_keypair ~url
+                              ~state_hash block_data ) ) ) ) )
+
+let start (mina : Mina_lib.t) =
+  let config = Mina_lib.config mina in
+  let transition_frontier = Mina_lib.transition_frontier mina in
+  match config.uptime_url with
+  | None ->
+      [%log' info config.logger] "Not running uptime service, no URL given" ;
+      ()
+  | Some url ->
+      [%log' info config.logger] "Starting uptime service using URL $url"
+        ~metadata:[("url", `String (Uri.to_string url))] ;
+      let slot_duration_ms =
+        Consensus.Configuration.t
+          ~constraint_constants:Genesis_constants.Constraint_constants.compiled
+          ~protocol_constants:Genesis_constants.compiled.protocol
+        |> Consensus.Configuration.slot_duration |> Float.of_int
+      in
+      let five_slots_span = Time.Span.of_ms (slot_duration_ms *. 5.0) in
+      (* every 5 slots, check whether block will be produced *)
+      Async.Clock.every' ~continue_on_error:true five_slots_span (fun () ->
+          [%log' trace config.logger]
+            "Uptime service determining which action to take" ;
+          let now = Time.now () in
+          let five_slots_from_now = Time.add now five_slots_span in
+          let get_next_producer_time_opt () =
+            match Mina_lib.next_producer_timing mina with
+            | None ->
+                [%log' trace config.logger]
+                  "Next producer timing not set for uptime service" ;
+                return None
+            | Some timing -> (
+                let open Daemon_rpcs.Types.Status.Next_producer_timing in
+                match timing.timing with
+                | Check_again _tm ->
+                    [%log' trace config.logger]
+                      "Next producer timing not available for uptime service" ;
+                    return None
+                | Produce prod_tm | Produce_now prod_tm ->
+                    return (Some (Block_time.to_time prod_tm.time)) )
+          in
+          match
+            ( config.gossip_net_params.addrs_and_ports.peer
+            , config.uptime_submitter_keypair )
+          with
+          | None, _ ->
+              [%log' warn config.logger]
+                "Daemon is not yet a peer in the gossip network, uptime \
+                 service not sending a produced block" ;
+              return ()
+          | Some _, None ->
+              (* should be unreachable *)
+              failwith
+                "No uptime submitter keypair, though a submitter URL was given"
```

Comment:
suggest "No uptime service submitter keypair..." -- thinking easiest to grep logs for "uptime service" -- almost all messages include that.